File: code/MatrixMult.java
import java.io.*;
import java.util.*;

public class MatrixMult {
    public static double[][] multiply(double[][] A, double[][] B, int n) {
        double[][] C = new double[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                double sum = 0.0;
                for (int k = 0; k < n; k++) {
                    sum += A[i][k] * B[k][j];
                }
                C[i][j] = sum;
            }
        }
        return C;
    }

    public static double[][] fillRandomMatrix(int n, Random rand) {
        double[][] mat = new double[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                mat[i][j] = rand.nextDouble();
            }
        }
        return mat;
    }

    public static void main(String[] args) {
        if (args.length != 2) {
            System.err.println("Uso: java MatrixMult <n> <repeticiones>");
            System.exit(1);
        }

        int n = Integer.parseInt(args[0]);
        int reps = Integer.parseInt(args[1]);
        Random rand = new Random(42); // Semilla fija

        double[][] A = fillRandomMatrix(n, rand);
        double[][] B = fillRandomMatrix(n, rand);

        double[] times = new double[reps];
        for (int r = 0; r < reps; r++) {
            long start = System.nanoTime();
            double[][] C = multiply(A, B, n);
            long end = System.nanoTime();
            double elapsed = (end - start) / 1_000_000_000.0; // nanoseg → seg
            times[r] = elapsed;
            System.out.printf("Rep %d: %.6f s%n", r + 1, elapsed);
        }

        double sum = 0.0;
        for (double t : times) sum += t;
        double avg = sum / reps;
        System.out.printf("Promedio: %.6f s%n", avg);

        // Guardar en CSV
        boolean fileExists = new File("results_java.csv").exists();
        try (PrintWriter pw = new PrintWriter(new FileWriter("results_java.csv", true))) {
            if (!fileExists) {
                pw.println("language,n,reps,avg_time");
            }
            pw.printf("Java,%d,%d,%.6f%n", n, reps, avg);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: code/matrix_mult.py
import time
import random
import sys
import csv
import os

def multiply_matrices(A, B, n):
    C = [[0.0] * n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            total = 0.0
            for k in range(n):
                total += A[i][k] * B[k][j]
            C[i][j] = total
    return C

def fill_random_matrix(n):
    return [[random.random() for _ in range(n)] for _ in range(n)]

def main():
    if len(sys.argv) != 3:
        print(f"Uso: {sys.argv[0]} <n> <repeticiones>")
        sys.exit(1)

    n = int(sys.argv[1])
    reps = int(sys.argv[2])
    random.seed(42)  # Reproducibilidad

    # Crear matrices (fuera del cronometraje)
    A = fill_random_matrix(n)
    B = fill_random_matrix(n)

    times = []
    for r in range(reps):
        start = time.perf_counter()
        C = multiply_matrices(A, B, n)
        end = time.perf_counter()
        elapsed = end - start
        times.append(elapsed)
        print(f"Rep {r + 1}: {elapsed:.6f} s")

    avg = sum(times) / len(times)
    print(f"Promedio: {avg:.6f} s")

    # Guardar en CSV
    file_exists = os.path.isfile('results_python.csv')
    with open('results_python.csv', 'a', newline='') as f:
        writer = csv.writer(f)
        if not file_exists:
            writer.writerow(['language', 'n', 'reps', 'avg_time'])
        writer.writerow(['Python', n, reps, f"{avg:.6f}"])

if __name__ == "__main__":
    main()